File: solutions/python/stack/car-fleet.py
# LeetCode 853 — Car Fleet (Medium)
# Category: Stack · Approach: Sort + sweep
# Time: O(n log n) | Space: O(n)
# Source: https://leetcode.com/problems/car-fleet/

def car_fleet(target: int, position: list[int], speed: list[int]) -> int:
    cars = sorted(zip(position, speed), reverse=True)
    fleets = 0
    cur = 0.0
    for pos, spd in cars:
        t = (target - pos) / spd
        if t > cur:
            fleets += 1
            cur = t
    return fleets

File: solutions/typescript/stack/car-fleet.ts
// LeetCode 853 — Car Fleet (Medium)
// Category: Stack · Approach: Sort + sweep
// Time: O(n log n) | Space: O(n)
// Source: https://leetcode.com/problems/car-fleet/

function carFleet(target: number, position: number[], speed: number[]): number {
  const idx = position.map((_, i) => i).sort((a, b) => position[b] - position[a]);
  let fleets = 0;
  let cur = 0;
  for (const i of idx) {
    const t = (target - position[i]) / speed[i];
    if (t > cur) {
      fleets++;
      cur = t;
    }
  }
  return fleets;
}
